File: Practica 2/C/ejercicios/ejercicio3.c
#include"ejercicio3.h"

void lecturaFichero3(){
    FILE *fp;
    fp = fopen("ficheros/PI2Ej3DatosEntrada.txt", "r");
    char numsLeidos[20];

    if(fp == NULL){
        printf("No se pudo abrir el fichero");
    }else{
        while(fgets(numsLeidos,20,fp)!=NULL){
            char * lineaAux = strtok(numsLeidos, ",");
            base = atol(lineaAux);
            lineaAux = strtok(NULL, ",");
            pot = atoi(lineaAux);
            printf("%ld %d\n", base, pot);
            printf("1. Iterativa (while): %ld\n", potenciaIterativa(base, pot));
            printf("2. Recursiva no final: %ld\n", potenciaRecNoFinal(base, pot));
            printf("3. Recursiva final: %ld\n\n", potenciaRecFinal(base, pot));

            
        }
    }
}

long potenciaIterativa(long i, int j){
    long res = 1;
    while(j>0){
        if(j%2==1){
            res*=i;
        }
        i*=i;
        j/=2;
    }
    return res;
}
long potenciaRecNoFinal(long i, int j){
    
    long res = 1;
    if(j<=0){
        return 1;
    }else{
        if(j%2==1){
            res *= i;
        }
        return res * potenciaRecFinal(i*i, j/2);
    }
}

long potenciaRecFinal(long i, int j){
    return potenciaRecFinalAux(i, j, 1);
}
long potenciaRecFinalAux(long i, int j, long res){
    if(j<=0){
        return res;
    }else{
        if(j%2==1){
            res *= i;
        }
        return potenciaRecFinalAux(i*i, j/2, res);
    }
}

File: Practica 2/Java/src/ejercicios/Ejercicio3.java
package ejercicios;

import java.io.File;
import java.io.FileNotFoundException;
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;
import java.util.stream.Stream;

import us.lsi.common.Tuple;
public class Ejercicio3 {

	
	public static void solucionEjercicio() {
		List<String> aux = new ArrayList<String>();
		aux = lecturaFichero();
		
		int i = 0;
		while(i<aux.size()) {
			String lineaActual[] = aux.get(i).split(",");
			Long a = Long.parseLong(lineaActual[0]);
			Integer b = Integer.parseInt(lineaActual[1]);
			System.out.println(aux.get(i));
			System.out.println("1. Iterativa (while): "+potenciaIterativa(a, b));
			System.out.println("2. Recursiva no final: "+potenciaRecNoFinal(a, b));
			System.out.println("3. Recursiva final: "+potenciaRecFinal(a, b));
			System.out.println("4. Funcional: "+potenciaFuncional(a, b));
			System.out.println();

			i++;
		}
	}
	
	private static List<String> lecturaFichero() {
		List<String> listaLineas = new ArrayList<String>();
		try {
			Scanner scanner = new Scanner(new File("ficheros/PI2Ej3DatosEntrada.txt"));
			
			while(scanner.hasNextLine()) {
				listaLineas.add(scanner.nextLine());
			}
			scanner.close();
			
		}catch(FileNotFoundException e) {
			e.printStackTrace();
		}
		return listaLineas;
	}
	
	private static Long potenciaIterativa(Long i, Integer j) {
		long res = 1;
	    while(j>0){
	        if(j%2==1){
	            res*=i;
	        }
	        i*=i;
	        j/=2;
	    }
	    return res;
	}
	
	private static Long potenciaRecFinal(Long i, Integer j) {
		return potenciaRecFinal(i, j, (long) 1);
	}
	private static Long potenciaRecFinal(Long i, Integer j, Long res) {
		if(j<=0) {
			return res;
		}else {
			if(j%2==1) {
				res*=i;
			}
			i*=i;
			j/=2;
			return potenciaRecFinal(i, j, res);
		}
	}
	
	private static Long potenciaRecNoFinal(Long i, Integer j) {
		long res = 1;
		if(j<=0) {
			return res;
		}else {
			if(j%2==1) {
				res*=i;
			}
			i*=i;
			j/=2;
			return res * potenciaRecNoFinal(i, j);
		}
	}
	
	public static Long potenciaFuncional(Long i, Integer j) {
		
		return Stream.iterate(Tuple.create(i, j, 1L), t->Tuple.create(t.v1*t.v1, t.v2/2, t.v2%2==1?t.v3*t.v1:t.v3))
				.dropWhile(t->t.v2>0).findFirst().get().v3;
		
	}
}
